cli/cli/helpers/grafloki/kubernetes_grafloki.go

```diff
@@ -0,0 +1,477 @@
+package grafloki
+
+import (
+	"context"
+	"fmt"
+	"github.com/kurtosis-tech/kurtosis/container-engine-lib/lib/backend_impls/kubernetes/kubernetes_manager"
+	"github.com/kurtosis-tech/stacktrace"
+	"github.com/sirupsen/logrus"
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	"time"
+)
+
+const (
+	grafanaServiceName                   = "kurtosis-grafana-service"
+	lokiServiceName                      = "kurtosis-loki-service"
+	grafanaDeploymentName                = "kurtosis-grafana-deployment"
+	lokiDeploymentName                   = "kurtosis-loki-deployment"
+	grafanaDatasourceConfigMapName       = "kurtosis-grafana-datasources"
+	graflokiNamespace                    = "kurtosis-grafloki"
+	grafanaNodePort                int32 = 30030
+	lokiNodePort                   int32 = 30031
+	lokiProbInitialDelaySeconds          = 5
+	lokiProbePeriodSeconds               = 10
+	lokiProbeTimeoutSeconds              = 10
+
+	// takes around 30 seconds for loki pod to become ready
+	lokiDeploymentMaxRetires    = 40
+	lokiDeploymentRetryInterval = 1 * time.Second
+)
+
+var lokiLabels = map[string]string{
+	lokiDeploymentName: "true",
+}
+var grafanaLabels = map[string]string{
+	grafanaDeploymentName: "true",
+}
+
+var httpApplicationProtocol = "http"
+
+func StartGrafLokiInKubernetes(ctx context.Context) (string, string, error) {
+	k8sManager, err := getKubernetesManager()
+	if err != nil {
+		return "", "", stacktrace.Propagate(err, "An error occurred getting Kubernetes Manager.")
+	}
+
+	var lokiHost string
+	doesGrafanaAndLokiExist, lokiHost := checkGrafanaAndLokiDeploymentExistence(ctx, k8sManager)
+	if err != nil {
+		return "", "", stacktrace.Propagate(err, "An error occurred checking if Grafana and Loki exist.")
+	}
+
+	if !doesGrafanaAndLokiExist {
+		lokiHost, err = createGrafanaAndLokiDeployments(ctx, k8sManager)
+		if err != nil {
+			return "", "", stacktrace.Propagate(err, "An error occurred creating Grafana and Loki deployments.")
+		}
+	}
+
+	logrus.Infof("Run `kubectl port-forward -n %v svc/%v %v:%v` to access Grafana service.", graflokiNamespace, grafanaServiceName, grafanaPort, grafanaNodePort)
+	return lokiHost, getGrafanaUrl(grafanaPort), nil
+}
+
+func createGrafanaAndLokiDeployments(ctx context.Context, k8sManager *kubernetes_manager.KubernetesManager) (string, error) {
+	glNamespace, err := k8sManager.CreateNamespace(ctx, graflokiNamespace, map[string]string{}, map[string]string{})
+	shouldRemoveNamespace := false
+	defer func() {
+		if shouldRemoveNamespace {
+			if err := k8sManager.RemoveNamespace(ctx, glNamespace); err != nil {
+				logrus.Warnf("Attempted to remove namespace '%v' after an error occurred creating it but an error occurred removing it.", graflokiNamespace)
+				logrus.Warnf("Manually remove namepsace with name: %v", glNamespace.Name)
+			}
+		}
+	}()
+	if err != nil {
```

Comment:
Same bug as comments from yesterday - error-checking is happening after the defer, which means the defer might be trying to delete something that never got created